src/util/src/tcache.c

```diff
@@ -19,688 +19,581 @@
+  if (pCacheObj->iterNum > 0) return;
+  if (pCacheObj->state != TAOS_CACHE_STATE_ACTIVE) return;
 
-  SHashTravSupp sup = {.pCacheObj = pCacheObj, .fp = fp, .time = time};
-  taosHashCondTraverse(pCacheObj->pHashTable, travHashTableFn, &sup);
-}
+  int32_t newSize = pCacheObj->size << 1;
+  if (newSize > TAOS_CACHE_MAX_SIZE) return;
 
-void* taosCacheTimedRefresh(void *handle) {
-  SCacheObj* pCacheObj = handle;
-  if (pCacheObj == NULL) {
-    uDebug("object is destroyed. no refresh retry");
-    return NULL;
-  }
+  void *p1 = realloc(pCacheObj->nodeList, newSize * sizeof(SCacheNode *));
+  if (p1 == NULL) return;
 
-  const int32_t SLEEP_DURATION = 500; //500 ms
-  int64_t totalTick = pCacheObj->refreshTime / SLEEP_DURATION;
-
-  int64_t count = 0;
-  while(1) {
-#if defined LINUX
-    usleep(500*1000);
-#else
-    taosMsleep(500);
-#endif
-
-    // check if current cache object will be deleted every 500ms.
-    if (pCacheObj->deleting) {
-      uDebug("%s refresh threads quit", pCacheObj->name);
-      break;
-    }
+  void *p2 = realloc(pCacheObj->nlock, newSize * sizeof(SRWLatch));
+  if (p2 == NULL) return;
 
-    if (++count < totalTick) {
-      continue;
-    }
+  int64_t st = taosGetTimestampUs();
 
-    // reset the count value
-    count = 0;
-    size_t elemInHash = taosHashGetSize(pCacheObj->pHashTable);
-    if (elemInHash + pCacheObj->numOfElemsInTrash == 0) {
-      continue;
-    }
+  pCacheObj->nodeList = (SCacheNode **)p1;
+  pCacheObj->nlock = (SRWLatch *)p2;
 
-    uDebug("%s refresh thread timed scan", pCacheObj->name);
-    pCacheObj->statistics.refreshCount++;
+  for (uint32_t hash=0; hash < pCacheObj->size; ++hash) {
+    SCacheNode *pNode = pCacheObj->nodeList[hash];
+    if (pNode == NULL) continue;
 
-    // refresh data in hash table
-    if (elemInHash > 0) {
-      int64_t now = taosGetTimestampMs();
-      doCacheRefresh(pCacheObj, now, NULL);
-    }
+    while (pNode) {
+      SCacheNode *pNext = pNode->next;
+      int nhash = (*pCacheObj->hashFp)(pNode->key, pNode->keyLen) % newSize;
+      if (hash != nhash) {
+        // remove from current list 
+        if (pNode->prev) {
+          pNode->prev->next = pNode->next;
+        } else {
+          pCacheObj->nodeList[hash] = pNode->next;
+        }
 
-    taosTrashcanEmpty(pCacheObj, false);
+        if (pNode->next) pNode->next->prev = pNode->prev;
+       
+        // add into new list
+        pNode->prev = NULL;
+        pNode->next = pCacheObj->nodeList[nhash];
+        if (pCacheObj->nodeList[nhash]) pCacheObj->nodeList[nhash] = pNode;
+      }
+
+      pNode = pNext;
+    }
   }
 
-  return NULL;
+  pCacheObj->size = newSize;
+
+  int64_t et = taosGetTimestampUs();
+  uDebug("cache:%s, hash table resized, new size:%d, elapsed time:%fms", pCacheObj->name, pCacheObj->size, (et - st)/1000.0); 
 }
+ 
+static void taosCacheResetObj(SCacheObj *pCacheObj) {
+  if (pCacheObj->state == TAOS_CACHE_STATE_EMPTY) return;
```

Comment:
seems there could be concurrent calls to this function, this check won't work in this case.